src/utils/xhr-utils.js

```diff
@@ -0,0 +1,624 @@
+    },
+  };
+
+  if (response['init']) {
+    const init = response['init'];
+    if (isArray(init.headers)) {
+      init.headers.forEach(entry => {
+        const headerName = entry[0];
+        const headerValue = entry[1];
+        lowercasedHeaders[String(headerName).toLowerCase()] =
+            String(headerValue);
+      });
+    }
+    if (init.status) {
+      data.status = parseInt(init.status, 10);
+    }
+    if (init.statusText) {
+      data.statusText = String(init.statusText);
+    }
+  }
+
+  if (isDocumentType) {
+    // TODO(prateekbh): remove this when integrating document fetcher.
+    data.responseXML =
+        new DOMParser().parseFromString(data.responseText, 'text/html');
+  }
+
+  return new FetchResponse(data);
+}
+
+/**
+ * Intercepts the XHR and proxies it through the viewer if necessary.
+ *
+ * XHRs are intercepted if all of the following are true:
+ * - The AMP doc is in single doc mode
+ * - The requested resource is not a 1p request.
+ * - The viewer has the `xhrInterceptor` capability
+ * - The Viewer is a trusted viewer or AMP is currently in developement mode
+ * - The AMP doc is opted-in for XHR interception (`<html>` tag has
+ *   `allow-xhr-interception` attribute)
+ *
+ * @param {!Window} win
+ * @param {?../service/ampdoc-impl.AmpDoc} ampdocSingle
+ * @param {string} input The URL of the XHR which may get intercepted.
+ * @param {!FetchInitDef} init The options of the XHR which may get
+ *     intercepted.
+ * @return {!Promise<!FetchResponse|!Response|undefined>}
+ *     A response returned by the interceptor if XHR is intercepted or
+ *     `Promise<undefined>` otherwise.
+ * @private
+ */
+export function getViewerInterceptResponse(win, ampdocSingle, input, init) {
+  if (!ampdocSingle) {
+    return Promise.resolve();
+  }
+  const viewer = Services.viewerForDoc(ampdocSingle);
+  const whenFirstVisible = viewer.whenFirstVisible();
+  if (isProxyOrigin(input) || !viewer.hasCapability('xhrInterceptor')) {
+    return whenFirstVisible;
+  }
+  const htmlElement = ampdocSingle.getRootNode().documentElement;
+  const docOptedIn = htmlElement.hasAttribute('allow-xhr-interception');
+  if (!docOptedIn) {
+    return whenFirstVisible;
+  }
+  return whenFirstVisible.then(() => {
+    return viewer.isTrustedViewer();
+  }).then(viewerTrusted => {
+    const isDevMode = getMode(win).development;
+    if (!viewerTrusted && !isDevMode) {
+      return;
+    }
+    const messagePayload = dict({
+      'originalRequest': toStructuredCloneable(input, init),
+    });
+    return viewer.sendMessageAwaitResponse('xhr', messagePayload)
+        .then(response =>
+          fromStructuredCloneable(response, init.responseType));
+  });
+}
+
+/**
+ * Setsup URL based on ampCors
+ * @param {!Window} win
+ * @param {string} input
+ * @param {!FetchInitDef} init The options of the XHR which may get
+ * intercepted.
+ */
+export function setupInput(win, input, init) {
```

Comment:
this is an atomic function to modify input, which was here: https://github.com/ampproject/amphtml/blob/master/src/service/xhr-impl.js#L375-L377

Reply:
this is an atomic function to modify input, which was here: https://github.com/ampproject/amphtml/blob/master/src/service/xhr-impl.js#L375-L377